Library/Homebrew/diagnostic.rb

```diff
@@ -842,6 +842,32 @@ def check_deleted_formula
         EOS
       end
 
+      def check_for_unnecessary_core_tap
+        return if ENV["HOMEBREW_DEVELOPER"]
+        return if ENV["HOMEBREW_NO_INSTALL_FROM_API"]
+        return unless CoreTap.instance.installed?
+
+        <<~EOS
+          You have an unnecessary local Core tap!
+          This can cause problems installing up-to-date formulae.
+          Please remove it by running:
+           brew untap #{CoreTap.instance.name}
+        EOS
+      end
+
+      def check_for_unnecessary_cask_tap
+        return if ENV["HOMEBREW_DEVELOPER"]
+        return if ENV["HOMEBREW_NO_INSTALL_FROM_API"]
+        return unless (cask_tap = Tap.fetch("homebrew", "cask")).installed?
```

Comment:
```suggestion
        cask_tap = Tap.fetch("homebrew", "cask")
        return unless cask_tap.installed?
```

at some point, not this PR, probably should just have a `CaskTap` or something class given how often we do this and how useful `CoreTap` is